x/operator/keeper/state_update.go

```diff
@@ -0,0 +1,494 @@
+			return err
+		}
+
+		// get the decimal of asset
+		assetInfo, err := k.restakingStateKeeper.GetStakingAssetInfo(ctx, assetID)
+		if err != nil {
+			return err
+		}
+		assetInfoRecord[assetID] = &AssetPriceAndDecimal{
+			Price:        price,
+			PriceDecimal: decimal,
+			Decimal:      assetInfo.AssetBasicInfo.Decimals,
+		}
+		assetUSDValue := CalculateShare(operatorAssetState.TotalAmountOrWantChangeValue, price, assetInfo.AssetBasicInfo.Decimals, decimal)
+		operatorUSDValue := CalculateShare(operatorAssetState.OperatorOwnAmountOrWantChangeValue, price, assetInfo.AssetBasicInfo.Decimals, decimal)
+		operatorOwnAssetUSDValue = operatorOwnAssetUSDValue.Add(operatorUSDValue)
+
+		// UpdateStateForAsset
+		changeState := types.OptedInAssetState{
+			Amount: operatorAssetState.TotalAmountOrWantChangeValue,
+			Value:  assetUSDValue,
+		}
+		err = k.UpdateStateForAsset(ctx, assetID, avsAddr, operatorAddress.String(), changeState)
+		if err != nil {
+			return err
+		}
+		totalAssetUSDValue = totalAssetUSDValue.Add(assetUSDValue)
+		assetFilter[assetID] = nil
+	}
+
+	// update the share value of operator itself, the input stakerID should be empty
+	err = k.UpdateStakerShare(ctx, avsAddr, "", operatorAddress.String(), operatorOwnAssetUSDValue)
+	if err != nil {
+		return err
+	}
+
+	// UpdateAVSShare
+	err = k.UpdateAVSShare(ctx, avsAddr, totalAssetUSDValue)
+	if err != nil {
+		return err
+	}
+	// UpdateOperatorShare
+	err = k.UpdateOperatorShare(ctx, avsAddr, operatorAddress.String(), totalAssetUSDValue)
+	if err != nil {
+		return err
+	}
+
+	// UpdateStakerShare
+	relatedAssetsState, err := k.delegationKeeper.DelegationStateByOperatorAssets(ctx, operatorAddress.String(), assetFilter)
+	if err != nil {
+		return err
+	}
+
+	for stakerID, assetState := range relatedAssetsState {
+		stakerAssetsUSDValue := sdkmath.LegacyNewDec(0)
+		for assetID, amount := range assetState {
+			singleAssetUSDValue := CalculateShare(amount.CanBeUndelegatedAmount, assetInfoRecord[assetID].Price, assetInfoRecord[assetID].Decimal, assetInfoRecord[assetID].PriceDecimal)
+			stakerAssetsUSDValue = stakerAssetsUSDValue.Add(singleAssetUSDValue)
+		}
+
+		err = k.UpdateStakerShare(ctx, avsAddr, stakerID, operatorAddress.String(), stakerAssetsUSDValue)
+		if err != nil {
+			return err
+		}
+	}
+
+	// update opted-in info
+	slashContract, err := k.avsKeeper.GetAvsSlashContract(ctx, avsAddr)
+	if err != nil {
+		return err
+	}
+	optedInfo := &types.OptedInfo{
+		SlashContract: slashContract,
+		// #nosec G701
+		OptedInHeight:  uint64(ctx.BlockHeight()),
+		OptedOutHeight: types.DefaultOptedOutHeight,
+	}
+	err = k.UpdateOptedInfo(ctx, operatorAddress.String(), avsAddr, optedInfo)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// OptOut call this function to opt out of AVS
+func (k *Keeper) OptOut(ctx sdk.Context, operatorAddress sdk.AccAddress, avsAddr string) error {
+	// check optedIn info
+	if !k.IsOptedIn(ctx, operatorAddress.String(), avsAddr) {
```

Comment:
In the private repo, we discussed as follows:

> This function's contents should not be triggered immediately on opt out. They should be done after the unbonding period on the AVS expires.

> Do you mean that the opt-out of AVS also requires a lock period? The undelegation of a staker definitely needs an unbonding period for slashing challenges. In the implementation, assets still need to wait a period to unbond when an undelegation is requested, even if the related operator has already opted out from the AVS. Therefore, the opt-out might not need an unbonding period.

You are correct. However, is there any data that you might wish to retain until the opt out start time + unbonding period finish ? For example, in the case of app chains / dogfood I am retaining the mapping of consensus public key to operator address. When slashing is triggered, the consensus key is supplied (by the SDK) as a parameter, from which I look up the operator address to slash. If I were to delete this data immediately upon opt out, such a look up would not be possible.

Reply:
It's to be determined whether the opt-out of AVS also requires a lock period.  In the current implementation of the operator module, the lock period of opt-out hasn't been considered. But I also noticed that the opt-out design in the dogfood has a lock period, this is an inconstancy that needs to be considered. If we don't set a lock period for the opt-out, about the state that will be deleted immediately when the AVS opts out, can we preserve it until the slash challenge period expires? That is to say, we can let the system delete the historical state according to the `PruningOptions`

Reply:
You are right that the dogfood module can use the historical context to determine the operator address using the consensus public key as an input, which can then be used for slashing.

However, in the undelegation hook, the code also uses the opt-out status of the operator (whether it is opting in or in the process of opting out) to determine the `unbondingCompletionEpoch`. This epoch is when the `DecrementUndelegationHoldCount` is called.

In other words, the undelegation matures either at the operator's opt out completion epoch, or the _usual_ undelegation completion epoch, whichever is earlier. I can store the former within the dogfood module and the latter can be calculated so I don't need an opt-out lock-in period implemented.

Reply:
@TimmyExogenous I have another question about this. If the operator opts in, waits for a while, opts out and then immediately opts back in, is this supported?

Reply:
The implemented `OptIn` and `OptOut` functionalities seem like they can support this. I didn't consider this situation when writing these codes. However, in the implementation, the related state will be updated immediately when the `OptIn` and `OptOut` are called. What potential issues might arise that are not supported?

Reply:
There are no immediate concerns that I can think of. Previously, I had used a queue implementation in the dogfood module which had made certain assumptions about the operator module (but this feature was still supported).